DHKeyAgreement.java

```diff
@@ -0,0 +1,127 @@
+import javax.crypto.*;
+import javax.crypto.spec.*;
+import javax.crypto.interfaces.*;
+import com.sun.crypto.provider.SunJCE;
+
+
+public class DHKeyAgreement{
+	private static final int keySize = 2048;
+	private static final String instance = "DH";
+
+	public DHKeyAgreement(){}
+
+	/* Since Client Inits contact, client chooses DH params */
+
+	public static KeyPair getClientKeyPair() throws NoSuchAlgorithmException, InvalidKeyException {
+		/* Client creates his/her own DH keypair with 2048 key size */
+
+		KeyPairGenerator clientKpairGen = KeyPairGenerator.getInstance("DH");
+		clientKpairGen.initialize(keySize);
+		KeyPair clientKpair;
+		try
+		{
+			clientKpair = clientKpairGen.generateKeyPair();
+			return clientKpair;
+		} catch (Exception e) {
+			throw new InvalidKeyException();
+		}
+	}
+
+	
+
+	public KeyPair getServerKeyPair(byte[] clientPubKeyEnc) throws NoSuchAlgorithmException, InvalidKeyException{
+
+		/* Server Instantiates DH public key from ClientEncoded value */
+		byte [] serverPubKeyEnc;
+		KeyFactory serverKeyFac = KeyFactory.getInstance("DH");
+		X509EncodedKeySpec x509KeySpec = new X509EncodedKeySpec(clientPubKeyEnc);
+		PublicKey clientPubKey;
+		
+		try{
+			clientPubKey = serverKeyFac.generatePublic(x509KeySpec);
+		}catch(Exception e){
+			throw new InvalidKeyException();
+		}
+
+		/* Server retrieves DH parameters used by Client (retrieved from clients pubkey) */
+
+		DHParameterSpec dhParamFromClientPubKey = ((DHPublicKey)clientPubKey).getParams();
+
+		/* Server generates own keypair using DH parameters from client public key*/
+
+		KeyPairGenerator serverKpairGen = KeyPairGenerator.getInstance("DH");
+		
+		try
+		{
+			serverKpairGen.initialize(dhParamFromClientPubKey);
+		}catch(Exception e){
+			throw new NoSuchAlgorithmException();
+		}
+
+		KeyPair serverKpair;// = serverKpairGen.generateKeyPair();
+		try
+		{
+			serverKpair = serverKpairGen.generateKeyPair();
+			return serverKpair;
+		} catch (Exception e) {
+			throw new InvalidKeyException();
+		}
+	}
+
+
+	public PublicKey getDHPublicKeyFromEncoded(byte[] pubKeyEncoded) throws NoSuchAlgorithmException, InvalidKeyException{
+		KeyFactory kf = KeyFactory.getInstance("DH");
+		X509EncodedKeySpec x509KeySpec = new X509EncodedKeySpec(pubKeyEncoded);
+		try
+		{
+			PublicKey dhPubKey = kf.generatePublic(x509KeySpec);
+			return dhPubKey;
+		} catch (Exception e) {
+			throw new InvalidKeyException();
+		}
+	}
+
+	 /*
+     * Converts a byte array to hex string
+     */
+    public static String toHexString(byte[] block) {
+        StringBuffer buf = new StringBuffer();
+        int len = block.length;
+        for (int i = 0; i < len; i++) {
+            byte2hex(block[i], buf);
+            if (i < len-1) {
+                buf.append(":");
+            }
+        }
+        return buf.toString();
+    }
+    /*
+     * Converts a byte to hex digit and writes to the supplied buffer
+     */
+    private static void byte2hex(byte b, StringBuffer buf) {
```

Comment:
Did you just use these byte2Hex & hex2Byte functions for debugging? If so, we should remove them.